packages/puppeteer-core/src/bidi/Target.ts

```diff
@@ -11,114 +11,92 @@ import {UnsupportedOperation} from '../common/Errors.js';
-import {type BrowsingContext, CdpSessionWrapper} from './BrowsingContext.js';
+import type {BrowsingContext} from './core/BrowsingContext.js';
 import {BidiPage} from './Page.js';
 
 /**
  * @internal
  */
-export abstract class BidiTarget extends Target {
-  protected _browserContext: BidiBrowserContext;
+export const targets = new WeakMap<object, Target>();
 
-  constructor(browserContext: BidiBrowserContext) {
+/**
+ * @internal
+ */
+
+export class BidiBrowserTarget extends Target {
+  readonly #browser: BidiBrowser;
+
+  constructor(browser: BidiBrowser) {
     super();
-    this._browserContext = browserContext;
-  }
+    targets.set(browser, this);
 
-  _setBrowserContext(browserContext: BidiBrowserContext): void {
-    this._browserContext = browserContext;
+    this.#browser = browser;
   }
 
   override asPage(): Promise<Page> {
     throw new UnsupportedOperation();
   }
-
-  override browser(): BidiBrowser {
-    return this._browserContext.browser();
-  }
-
-  override browserContext(): BidiBrowserContext {
-    return this._browserContext;
-  }
-
-  override opener(): never {
-    throw new UnsupportedOperation();
+  override url(): string {
+    return '';
   }
-
   override createCDPSession(): Promise<CDPSession> {
     throw new UnsupportedOperation();
   }
-}
-
-/**
- * @internal
- */
-export class BiDiBrowserTarget extends BidiTarget {
-  override url(): string {
-    return '';
-  }
-
   override type(): TargetType {
     return TargetType.BROWSER;
   }
+  override browser(): BidiBrowser {
+    return this.#browser;
+  }
+  override browserContext(): BidiBrowserContext {
+    return this.#browser.defaultBrowserContext();
+  }
+  override opener(): Target | undefined {
+    return undefined;
+  }
 }
 
 /**
  * @internal
  */
-export class BiDiBrowsingContextTarget extends BidiTarget {
-  protected _browsingContext: BrowsingContext;
+
+export class BidiPageTarget extends Target {
+  readonly #context: BidiBrowserContext;
+  readonly #page: Promise<BidiPage>;
+
+  _url = '';
 
   constructor(
-    browserContext: BidiBrowserContext,
-    browsingContext: BrowsingContext
+    browsingContext: Promise<BrowsingContext>,
+    context: BidiBrowserContext
   ) {
-    super(browserContext);
-
-    this._browsingContext = browsingContext;
+    super();
+    this.#context = context;
+    this.#page = browsingContext.then(context => {
+      return BidiPage.create(context, this);
+    });
   }
 
+  override asPage(): Promise<BidiPage> {
+    return this.#page;
+  }
   override url(): string {
-    return this._browsingContext.url;
+    return this._url;
   }
-
   override async createCDPSession(): Promise<CDPSession> {
-    const {sessionId} = await this._browsingContext.cdpSession.send(
-      'Target.attachToTarget',
-      {
-        targetId: this._browsingContext.id,
-        flatten: true,
-      }
-    );
-    return new CdpSessionWrapper(this._browsingContext, sessionId);
+    const page = await this.asPage();
```

Comment:
This does not seem correct, you should be able to create session on other targets as well.